### ISSUE TITLE
Broken osx backend

### ISSUE THREAD

**Issue by beckermr** [conda-forge/matplotlib-feedstock#44]
I got this error on osx.

``` python
Python 3.5.2 | packaged by conda-forge | (default, Jul  7 2016, 12:53:03) 
Type "copyright", "credits" or "license" for more information.

IPython 5.0.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.
[TerminalIPythonApp] WARNING | Eventloop or matplotlib integration failed. Is matplotlib installed?
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/IPython/core/shellapp.py in <lambda>(key)
    196         shell = self.shell
    197         if self.pylab:
--> 198             enable = lambda key: shell.enable_pylab(key, import_all=self.pylab_import_all)
    199             key = self.pylab
    200         elif self.matplotlib:

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/IPython/core/interactiveshell.py in enable_pylab(self, gui, import_all, welcome_message)
   2972         from IPython.core.pylabtools import import_pylab
   2973 
-> 2974         gui, backend = self.enable_matplotlib(gui)
   2975 
   2976         # We want to prevent the loading of pylab to pollute the user's

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/IPython/core/interactiveshell.py in enable_matplotlib(self, gui)
   2933                 gui, backend = pt.find_gui_and_backend(self.pylab_gui_select)
   2934 
-> 2935         pt.activate_matplotlib(backend)
   2936         pt.configure_inline_support(self, backend)
   2937 

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/IPython/core/pylabtools.py in activate_matplotlib(backend)
    293     matplotlib.rcParams['backend'] = backend
    294 
--> 295     import matplotlib.pyplot
    296     matplotlib.pyplot.switch_backend(backend)
    297 

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/matplotlib/pyplot.py in <module>()
    112 
    113 from matplotlib.backends import pylab_setup
--> 114 _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
    115 
    116 _IP_REGISTERED = None

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/matplotlib/backends/__init__.py in pylab_setup()
     30     # imports. 0 means only perform absolute imports.
     31     backend_mod = __import__(backend_name,
---> 32                              globals(),locals(),[backend_name],0)
     33 
     34     # Things we pull in from all backends

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/matplotlib/backends/backend_macosx.py in <module>()
     22 
     23 import matplotlib
---> 24 from matplotlib.backends import _macosx
     25 
     26 

ImportError: cannot import name '_macosx'
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-ed9806ce3570> in <module>()
----> 1 import seaborn as sns

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/seaborn/__init__.py in <module>()
----> 1 from .rcmod import *
      2 from .utils import *
      3 from .palettes import *
      4 from .linearmodels import *
      5 from .categorical import *

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/seaborn/rcmod.py in <module>()
      6 import matplotlib as mpl
      7 
----> 8 from . import palettes
      9 
     10 mpl_ge_150 = LooseVersion(mpl.__version__) >= '1.5.0'

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/seaborn/palettes.py in <module>()
     10 from .external.six.moves import range
     11 
---> 12 from .utils import desaturate, set_hls_values, get_color_cycle
     13 from .xkcd_rgb import xkcd_rgb
     14 from .crayons import crayons

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/seaborn/utils.py in <module>()
     10 import matplotlib as mpl
     11 import matplotlib.colors as mplcol
---> 12 import matplotlib.pyplot as plt
     13 
     14 from distutils.version import LooseVersion

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/matplotlib/pyplot.py in <module>()
    112 
    113 from matplotlib.backends import pylab_setup
--> 114 _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
    115 
    116 _IP_REGISTERED = None

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/matplotlib/backends/__init__.py in pylab_setup()
     30     # imports. 0 means only perform absolute imports.
     31     backend_mod = __import__(backend_name,
---> 32                              globals(),locals(),[backend_name],0)
     33 
     34     # Things we pull in from all backends

/Users/mbecker/miniconda3/envs/civis/lib/python3.5/site-packages/matplotlib/backends/backend_macosx.py in <module>()
     22 
     23 import matplotlib
---> 24 from matplotlib.backends import _macosx
     25 
     26 

ImportError: cannot import name '_macosx'
```

**Comment by beckermr** [comment]
Some more info:

```             platform : osx-64
        conda version : 4.0.6
  conda-build version : not installed
       python version : 3.5.1.final.0
     requests version : 2.9.1
     root environment : /Users/mbecker/miniconda3  (writable)
  default environment : /Users/mbecker/miniconda3/envs/civis
     envs directories : /Users/mbecker/miniconda3/envs
        package cache : /Users/mbecker/miniconda3/pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/osx-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/osx-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : /Users/mbecker/.condarc
    is foreign system : False
```

I rolled back this install which is why the python version differs from above.

**Comment by astrofrog** [comment]
I'm seeing this too with 2.0.0b3:

```
$ python -c 'import matplotlib; matplotlib.use("MacOSX"); import matplotlib.pyplot'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/matplotlib/pyplot.py", line 114, in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
  File "/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/matplotlib/backends/__init__.py", line 32, in pylab_setup
    globals(),locals(),[backend_name],0)
  File "/Users/tom/miniconda3/envs/dev35/lib/python3.5/site-packages/matplotlib/backends/backend_macosx.py", line 19, in <module>
    from matplotlib.backends import _macosx
ImportError: cannot import name '_macosx'
```

```
             platform : osx-64
        conda version : 4.0.5
  conda-build version : not installed
       python version : 3.5.1.final.0
     requests version : 2.9.1
     root environment : /Users/tom/miniconda3  (writable)
  default environment : /Users/tom/miniconda3/envs/dev35
     envs directories : /Users/tom/miniconda3/envs
        package cache : /Users/tom/miniconda3/pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/osx-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/osx-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : None
    is foreign system : False
```

**Comment by ocefpaf** [comment]
@jjhelmus or @pelson can you test that? No Mac here on my side :unamused:

**Comment by aphearin** [comment]
@beckermr and @astrofrog - did either of you ever resolve this issue? I just started using conda-forge and this error came up for me as well. 

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/aphearin/anaconda/lib/python2.7/site-packages/matplotlib/pyplot.py", line 114, in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
  File "/Users/aphearin/anaconda/lib/python2.7/site-packages/matplotlib/backends/__init__.py", line 32, in pylab_setup
    globals(),locals(),[backend_name],0)
  File "/Users/aphearin/anaconda/lib/python2.7/site-packages/matplotlib/backends/backend_macosx.py", line 24, in <module>
    from matplotlib.backends import _macosx
ImportError: cannot import name _macosx
```

```
             platform : osx-64
        conda version : 4.1.12
    conda-env version : 2.5.2
  conda-build version : 2.0.6
       python version : 2.7.12.final.0
     requests version : 2.11.1
     root environment : /Users/aphearin/anaconda  (writable)
  default environment : /Users/aphearin/anaconda
     envs directories : /Users/aphearin/anaconda/envs
        package cache : /Users/aphearin/anaconda/pkgs
         channel URLs : https://conda.anaconda.org/conda-forge/osx-64/
                        https://conda.anaconda.org/conda-forge/noarch/
                        https://repo.continuum.io/pkgs/free/osx-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/osx-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : /Users/aphearin/.condarc
         offline mode : False
    is foreign system : False

```

**Comment by beckermr** [comment]
@aphearin I know of one other person with this error. My solution was not to use conda-forge. :(

**Comment by aphearin** [comment]
Any major side-effects I should be aware of with this bug? I seem to have no trouble using matplotlib plots in jupyter notebooks to generate and save my figures.

**Comment by beckermr** [comment]
No idea. I didn't even get that far. Once my backend was broken on the command line, I gave up.

**Comment by astrofrog** [comment]
I haven't run into that error recently

**Comment by nirum** [comment]
I just ran into the same error. Ended up not using conda-forge

**Comment by tacaswell** [comment]
@beckermr Suggest using one of the qt-based backends instead.

**Comment by aphearin** [comment]
That's what I did, @tacaswell, but I wish this wasn't necessary, it also made me consider the same option @beckermr did.

**Comment by dopplershift** [comment]
On one hand I get it--on the other hand I'm not sure what's wrong with using Qt4Agg. It's probably the most feature-filled of all the backends.

**Comment by beckermr** [comment]
(Hey @tacaswell! I hope all is well! I am back in Chicago if you happen to swing this way.)

IMHO, this is a very typical dev response to an actual problem. Just use a feature that is not broken right now. Sure I get it. However, this part of matplotlib installs from conda-forge appears to actually be broken last I checked and it is not from conda defaults or pip. Clearly something needs to be fixed, right?

**Comment by dopplershift** [comment]
Yes, it is a typical dev response. 😈   But I disagree about what's broken here. It'd be nice if conda-forge could successfully build the osx backend (and it'd be good to solve the build problem), but they use a supported method to exclude the backend from the built package. IMO, we (matplotlib) should fallback nicely to another backend in the case that the requested backend isn't actually available.

TLDR; Given that Matplotlib supports build-time selection of backends, it should have better behavior when users (especially in a config file) ask for something that wasn't built.

**Comment by beckermr** [comment]
Ohh yes. I agree this is a conda-forge problem, not matplotlib. I just want to be clear that is a problem, and not just a user issue.

**Comment by dopplershift** [comment]
Actually I was more arguing that it's a matplotlib user experience, not a conda-forge problem. At any rate there is a problem.

**Comment by aphearin** [comment]
I think the point that @beckermr is trying to make is just this: 1. other users switching to conda-forge are going to encounter this problem, 2. many such users will be in our position and rely on their IPython/matplotlib environment for their daily work, 3. it is entirely reasonable for such users to at least balk at the need to change their backend. 

I fully agree with @dopplershift, and in fact I voted with my feet since I still use conda-forge and I did switch my backend after all. And while the ultimate solution may come from matplotlib changing their backend fallback, all the same I do think that one way or another this *is* a problem for conda-forge to take seriously since it may scare off some of the desired user base.

**Comment by astrofrog** [comment]
I for one like using the OS X backend because it supports retina displays well (Qt4 doesn't, and Qt5 is still a work in progress). I really think the conda-forge build should support the MacOS X backend.

**Comment by ocefpaf** [comment]
> I really think the conda-forge build should support the MacOS X backend.

I agree. The problem is to find Mac owners volunteers to fix the recipe. I do not own a Mac and I have no idea of what needs to be done there.

**Comment by dopplershift** [comment]
Ah--I wasn't aware of that benefit of the osx backend.

And to be clear, I'm not arguing against having it--just it's not "broken" when a backend is not present due to the configuration in `setup.cfg`. Matplotlib needs a better runtime experience around packages that are configured as such.

**Comment by dopplershift** [comment]
So I built the package locally, enabled the macosx backend in the build...and everything worked fine. macOS 10.12.1 (Sierra) here.

**Comment by ocefpaf** [comment]
> So I built the package locally, enabled the macosx backend in the build...and everything worked fine. macOS 10.12.1 (Sierra) here.

Awesome. All we need is a patch that applied the proper config flag when `Darwin` is detected in the same fashion as https://github.com/conda-forge/matplotlib-feedstock/blob/master/recipe/cfg_qt4agg.patch

However, that will make the `macosx` backend the default. Ideally we should break matplotlib into small backend packages. But that is a long term plan. For now, if people agree, changing the default from `Qt4` to `macosx` on `OS X` seems OK to me (caveat: I am not a mac user).

**Comment by dopplershift** [comment]
I've taken out the part disabling the build in #87. Let's see whether that succeeds. If it does, we can at least have packages that have it available and will work for people with `'macosx'` set as the backend in `matplotlibrc`.

**Comment by michaelaye** [comment]
@tacaswell one problem i currently have with the Qt backends is that they create icons in the Dock.app when loaded. This is highly annoying when running an ipcluster with 8 cores (I have no clue if it's possible to have ipcluster load a different backend than my interactive enviroment, that could be a solution.)

**Comment by michaelaye** [comment]
@ocefpaf i take it that's the default taken when no matplotlibrc file exists? And if it does, then that setting would be used, correct?

**Comment by tacaswell** [comment]
@michaelaye You probably want to set `matplotlib.rcParams['backend'] = 'Agg'` on your cluster workers before you send them anything which will import pyplot.  Depending on what you are passing back and forth to your works that will either just work or fail with really weird errors :smiling_imp: .

I have no idea why the qt backends try to throw an icon in your doc.  Is that something mpl is doing intentionally (if so we should stop) or is that a Qt thing (in which case _they_ should stop).

**Comment by tacaswell** [comment]
in either case this should be fixed by #89 and #87 so closing.

Please ping to have this re-opened if the new packages are still broken.

**Comment by kmuehlbauer** [comment]
see https://github.com/conda-forge/wradlib-feedstock/pull/20#issuecomment-268952362

**Comment by ocefpaf** [comment]
Would a patch adding  the macos backend as the default suffice to fix this?
I don't have a mac to test it :-(

On Dec 29, 2016 9:29 AM, "Thomas Robitaille" <notifications@github.com>
wrote:

@tacaswell <https://github.com/tacaswell> @ocefpaf
<https://github.com/ocefpaf> - this is still broken for me, and in fact
this now breaks with a clean install:

$ python -c 'import matplotlib.pyplot as plt'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/tom/miniconda3/envs/debug-mpl-1/lib/python3.5/site-packages/matplotlib/pyplot.py",
line 114, in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
  File "/Users/tom/miniconda3/envs/debug-mpl-1/lib/python3.5/site-packages/matplotlib/backends/__init__.py",
line 32, in pylab_setup
    globals(),locals(),[backend_name],0)
  File "/Users/tom/miniconda3/envs/debug-mpl-1/lib/python3.5/site-packages/matplotlib/backends/backend_macosx.py",
line 24, in <module>
    from matplotlib.backends import _macosx
RuntimeError: Python is not installed as a framework. The Mac OS X
backend will not be able to function correctly if Python is not
installed as a framework. See the Python documentation for more
information on installing Python as a framework on Mac OS X. Please
either reinstall Python as a framework, or try one of the other
backends. If you are Working with Matplotlib in a virtual enviroment
see 'Working with Matplotlib in Virtual environments' in the
Matplotlib FAQ

My environment is:

$ conda list
# packages in environment at /Users/tom/miniconda3/envs/debug-mpl-1:
#
ca-certificates           2016.9.26                     0    conda-forge
certifi                   2016.9.26                py35_0    conda-forge
cycler                    0.10.0                   py35_0    conda-forge
freetype                  2.6.3                         1    conda-forge
libpng                    1.6.26                        0    conda-forge
matplotlib                1.5.3               np111py35_4    conda-forge
mkl                       11.3.3                        0
ncurses                   5.9                          10    conda-forge
numpy                     1.11.2                   py35_0
openssl                   1.0.2h                        3    conda-forge
pip                       9.0.1                    py35_0    conda-forge
pyparsing                 2.1.10                   py35_0    conda-forge
python                    3.5.2                         3    conda-forge
python-dateutil           2.6.0                    py35_0    conda-forge
pytz                      2016.10                  py35_0    conda-forge
readline                  6.2                           0    conda-forge
setuptools                32.3.0                   py35_0    conda-forge
six                       1.10.0                   py35_0    conda-forge
sqlite                    3.13.0                        1    conda-forge
tk                        8.5.19                        1    conda-forge
wheel                     0.29.0                   py35_0    conda-forge
xz                        5.2.2                         0    conda-forge
zlib                      1.2.8                         3    conda-forge

This is causing issues on Travis with MacOS X for packages where I use
Matplotlib from conda-forge.

Note that to get the above error you need to not have any matplotlibrc file
present. If you have a completely empty matplotlibrc file, the crash
doesn't happen (@tacaswell <https://github.com/tacaswell>, is that
expected?)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/conda-forge/matplotlib-feedstock/issues/44#issuecomment-269623650>,
or mute the thread

**Comment by dopplershift** [comment]
I'm not convinced there's anything wrong with the conda-forge package. The error message mentions needing a framework, which with conda I've always had to use with `pythonw`. Have you been able to use the osx backend without using `pythonw` on conda?

**Comment by michaelaye** [comment]
Yes, whenever I install the MPL package from conda defaults channel, it just works. I just did it to reconfirm:
```bash
conda remove matplotlib --force
conda install --override-channels -c defaults matplotlib
ipython (which imports mpl in its config)
-> works
conda update --all (picking up conda-forge's mpl again
ipython
-> above RunTimeError
```

**Comment by tacaswell** [comment]
@astrofrog re matplotlibrc, yes that will cause it to fall back to the default defaults (for lack of a better phrase) which uses Agg rather than the defaults set at build time for the default backend (which is OSX in the mac case).

**Comment by dopplershift** [comment]
Alright this is weird. The following command (using Python from the defaults channel):
```sh
python simple_plot.py -dmacosx
```
works fine with matplotlib 1.5.3 from defaults (including a nice plot window), but gives me the `RuntimeError` (about a framework build) with matplotlib 1.5.3 from conda-forge. (`pythonw` works for both).

@tacaswell Does that make any sense to you whatsoever? I have trouble understanding how that exception could depend on how we build matplotlib. The only theory I can think of is Continuum has some other magic taking place in their build...

**Comment by dopplershift** [comment]
Ah, there's a [patch](https://github.com/ContinuumIO/anaconda-recipes/blob/master/matplotlib/osx-frame.patch) over in ContinuumIO/anaconda-recipes. 

I'm inclined to just incorporate that patch here and see about fixing that check in matplotlib proper.
@tacaswell do you have any thoughts?

**Comment by dopplershift** [comment]
See #92 .

**Comment by astrofrog** [comment]
+1 to just incorporating the patch
